iree/vm/stack.c

```diff
@@ -16,87 +16,706 @@
+//
+// Calling convention
+// ------------------
+// Stack frames are managed by callees. A caller will provide a list of argument
+// and result registers in the caller frame that should be passed into and out
+// of the callee. The callee will first push its stack frame and reserve the
+// register storage it requires, copy or move the argument values from the
+// caller frame into the callee frame, and then begin execution. Upon return the
+// callee function will move return registers from its frame out to the caller
+// frame and pop its stack frame.
+//
+// By making the actual stack frame setup and teardown callee-controlled we can
+// have optimized implementations that treat register storage differently across
+// various frames. For example, native modules that store their registers in
+// host-machine specific registers can marshal the caller registers in/out of
+// the host registers (or stack/etc) without exposing the actual implementation
+// to the caller.
+//
+// Calling into the VM
+// -------------------
+// Calls from external code into the VM such as via iree_vm_invoke reuse the
+// same calling convention as internal-to-internal calls: callees load arguments
+// from the caller frame and store results into the caller frame. To enable this
+// a set of iree_vm_stack_external_enter/iree_vm_stack_external_leave functions
+// are provided to setup the special `[external]` stack frame that allows the
+// external code to marshal in the arguments and marshal out the results.
+//
+// Marshaling arguments is easy given that the caller controls these and we can
+// trivially map the ordered set of argument types into the ABI registers.
+//
+// TODO(#1979): know result types ahead of time.
+// Results are more difficult as today we do not know the order or type of the
+// function results upon entry. As such when the callee goes to leave its frame
+// and store the results back to the caller we cannot place directly into
+// registers that then the iree_vm_stack_external_leave can access. To work
+// around this until we can more cleanly provide this information we check for
+// external callers when leaving frames and duplicate the register mapping. This
+// can result in use-after-free if the result register mapping was not in rdata
+// however today both our C++ and bytecode implementations are.
+//
+// A side-effect (beyond code reuse) is that ref types are retained by the VM
+// for the entire lifetime they may be accessible by VM routines. This lets us
+// get rich stack traces without needing to hook into external code and lets us
+// timeshift via coroutines where we may otherwise not know when the external
+// caller will resume a yielded call and actually read back the results.
+//
+// The overhead of this marshaling is minimal as external functions can always
+// use move semantics on the ref objects. Since we are reusing the normal VM
+// code paths which are likely still in I$ the bulk of the work amounts to some
```

Comment:
What's `I$`?

Reply:
instruction cache ($ is shorthand for cache, so you'll often see "I$"/"D$" around --- changing to be clearer